[2단계 - 장바구니 협업] 돔하디(이동현) 미션 제출합니다.

DomMorello opened the pull request:
> 안녕하세요 오스틴님 :) 장바구니 협업 미션 스텝 2 제출합니다. 백엔드 크루들과의 긴밀한 협의를 통해서 API 명세를 설계하는 과정에서 서버 개발자들이 무슨 일을 하는지 어렴풋이나마 이해할 수 있었고 서로의 입장을 공유할 수 있는 좋은 기회였습니다 ㅎㅎㅎ 스텝 1에서 API 요청을 mock한 부분이 실제 기능과 다르게 무조건 성공하는 로직이었다보니 기능이 정상적으로 동작하는지 점검이 어려우셨을텐데 이번에는 실제 서버에서 동작하니 기능 점검을 원활하게 하실 수 있을겁니다! ㅎㅎㅎㅎ피드백 주시면 감사하겠습니다.

## 로컬 실행 방법
```bash
npm install
npm run start
npm run storybook
```
> 실제 배포된 백엔드 서버와 연동되어있기 때문에 클라이언트까지 배포하지는 못했습니다 :( 로컬 환경에서 테스트해주세요 ㅎㅎ

## step2 기능 요구사항
- [x] 로컬 기반의 장바구니를 서버 기반으로 변경
- [x] API 호스트를 변경해 모든 백엔드 크루들의 API를 테스트할 수 있어야한다.

> `src/constants/index.js` 파일에 `BASE_SERVER_URL` 상수가 정의돼있습니다. 백엔드 크루들이 구현한 서버가 각각 배포되어있는데 서버의 URL 을 바꿔가면서 점검해볼 수 있도록 해놨습니다. 같은 API 명세를 공유하기 때문에 서버에 상관없이 같은 동작을 해야 합니다.

## 고민한 부분
### 초기 라우팅 개선
env의 `PUBLIC_URL` 을 활용하여 기본 메인 페이지가 처음 url로 동작하도록 수정했습니다.

### Redirect를 선언적인 방법으로 수정
react router v6 에서 `useNaviagte`를 활용하여 특정 상황에서 라우팅을 시켜주는 방식을 사용했었는데 보통 이런 방식들이 useEffect 훅 안에서 진행된 부분이 있었습니다. 크게 달라지는 점은 없겠지만 useEffect 훅이 개인적으로 가독성이 좋지 않고 개발자의 의도와 상관없이 버그를 일으킬 가능성이 어느 정도 있다고 생각해서 선언적으로 하는 방법으로 수정했습니다. 이로써 리다이렉트를 위해서 존재하는 useEffect 를 대체할 수 있었습니다.

```js
  //  LoginPage/index.jsx
  return isLoggedIn ? (
    <Navigate to={ROUTES.ROOT} replace />
  ) : (
    <LoginPageContainer>
      <PageHeader>로그인</PageHeader>
      ...
    </LoginPageContainer>
  );
```
### 리덕스 스토어 단점
장바구니 목록을 redux store에서 관리하고 있습니다. 그런데 현재 기능상으로는 DB에 어차피 장바구니 목록이 저장돼있습니다. 모든 페이지를 방문할때마다 서버로 장바구니 목록을 요청해서 받아오는데 이를 또 redux store에 업데이트를 해줍니다. 이런 과정에서 어떤 기능이 수정되거나 할 때 DB와 redux store가 일관되게 유지해야 하는 번거로움이 있습니다. 이는 관리포인트가 하나 더 느는 것 같고 이렇게 함으로써 얻는 효용이 없다는 생각을 했습니다.

### token, userId 쿠키에 저장
로그인을 할 때 서버에서 보내주는 accessToken과 userId를 브라우저의 쿠키에 저장해서 관리했습니다. 이를 통해서 redux store의 `isLoggedIn` 상태로 관리하였고 이 상태를 통해서 로그인 여부를 판단하고 행동을 결정(리다이렉트 등)합니다. 쿠키에 저장할 때 서버에서 보내주는 `expirationTime` 을 계산해서 해당 시간에 맞게 만료되도록 설정했습니다. 이렇게 함으로써 토큰이 만료된 경우 사용자는 다시 로그인해야 한다는 사실을 인지할 수 있을 것입니다. (에러메세지: 장바구니를 추가하려면 로그인을 해야 합니다 등)

> 피드백 주시면 잘 생각해보고 적용하겠습니다! 감사합니다 :)

DomMorello committed:
fix: 초기 라우팅 수정

DomMorello committed:
docs: 추가된 기능 구현 목록 작성

DomMorello committed:
feat: 로그인 여부에 따라 nav버튼 UI 변경

DomMorello committed:
refactor: nickname 네이밍을 서버와 동일하게 username으로 수정

DomMorello committed:
feat: accessToken, id 쿠키에 저장 및 활용

- 로그인 시에 쿠키에 token, id 저장
- redux store에 isLoggedIn state 추가 및 accessToken 제거

DomMorello committed:
feat: 헤더 로그아웃시 쿠키 제거

DomMorello committed:
feat: 장바구니 페이지 로그인하지 않은 유저 리다이렉트 구현

DomMorello committed:
feat: 회원가입 페이지 로그인한 유저 리다이렉트 구현

DomMorello committed:
refactor: 리다이렉트 state이용하여 수정

DomMorello committed:
feat: 수정된 서버 API 명세에 맞게 수정

id -> productId 로 이름 변경
서버요청 url 구조 변경

DomMorello committed:
chore: msw 함수 주석 처리

DomMorello committed:
feat: 재고를 추가해서 수량을 증가하면 경고함

DomMorello committed:
feat: 로그인 하지 않은 유저 장바구니 추가 제한

DomMorello committed:
feat: 재고 에러 발생시 서버에서 주는 에러메세지 보여주기

DomMorello committed:
fix: 새로고침시 user 정보 초기화되는 버그 수정

App 에서 새로고침마다 user정보를 가져옴
가져온 정보를 redux에 업데이트함
이를 위해 store로 App을 감쌈

DomMorello committed:
chore: server 선택 사항 추가

DomMorello committed:
fix: storybook import 에러 수정

austinpark420 commented on src/index.js:
```diff
@@ -1,22 +1,47 @@
-  const { worker } = require("./mocks/browser");
-  worker.start();
-}
+const main = () => {
```
main 함수 아래 주석으로 처리되어 있던데 개발할 때 사용하시는건가요?

DomMorello replied:
이 부분은 msw를 이용해서 할 때 주석을 풀어서 하는 코드입니다. 그런데 step2 는 실제 서버와 연동하는 테스트이기 때문에 주석처리를 해 놓은 상태입니다.

austinpark420 commented on src/components/App.jsx:
```diff
@@ -1,14 +1,10 @@
-import { applyMiddleware, createStore } from "redux";
-import { composeWithDevTools } from "@redux-devtools/extension";
-import ReduxThunk from "redux-thunk";
 import { Route, Routes } from "react-router-dom";
```
App.tsx를 components 디렉토리 밑에 위치시킨 이유가 있을까요?

DomMorello replied:
모든 page들을 포함하는 가장 상위의 컴포넌트라고 생각했습니다!

austinpark420 commented on src/reducers/user.js:
```diff
@@ -98,39 +134,53 @@ const reducer = (state = initialState, action) => {
+        isLoggedIn: state.isLoggedIn,
         isLoading: false,
         data: state.data,
         errorMessage: "",
```
login 함수에서 메세지가 있을때 에러를 띄어주는게 BE에서 그렇게 보내주는 방법밖에 없다고 한건가요....?
```javascript
    const data = await response.json();
    if (data.message) {
      throw new Error(data.message);
    }
```

DomMorello replied:
질문을 제대로 이해하지 못했습니다.. 서버에서 message가 있으면 그건 에러가 발생한 것이기 때문에 에러를 발생시킵니다. 에러가 발생하면 dispatch를 통해서 redux store의 isLoggedIn 의 상태가 변경됩니다.

austinpark420 replied:
`message가 있으면 그건 에러가 발생한 것이기 때문에 에러를 발생시킵니다` 라는 히스토리를 모른다면 다른 개발자들이 코드를 봤을 때 바로 코드를 이해하기 어려운 구조인것 같습니다. 다른 프로젝트에서는 BE에게 message외 다른 플래그값을 전달받는건 어떨까요?

austinpark420 commented on src/components/layout/Header/index.jsx:
```diff
@@ -10,23 +11,25 @@ import { UnderlinedButton } from "./NavButton/styled";
   const navigator = useNavigate();
 
   const logout = () => {
     dispatch({ type: USER_ACTION.LOGOUT });
-    navigator(ROUTES.ROOT);
+    deleteCookie(COOKIE_KEY.TOKEN);
+    deleteCookie(COOKIE_KEY.USER_ID);
+    navigator(ROUTES.ROOT, { replace: true });
   };
```
logout 함수는 reducers/user.js 에서 관리 안하는 이유가 있을까요?

DomMorello replied:
logout 함수를 보면 `dispatch({ type: USER_ACTION.LOGOUT });` 를 통해서 store의 isLoggedIn 의 상태를 false 로 변화시킵니다. 

```js
const reducer = (state = initialState, action) => {
  ...
    case USER_ACTION.LOGOUT:
      return { ...initialState, isLoggedIn: false };
  ...
}
```
관리를 하고 있다고 생각하는데 리뷰어님은 어떤 부분을 보시고 관리되지 않는다고 생각하셨는지 궁금합니다 ㅎㅎ

austinpark420 replied:
앗 이 부분 제가 잘못봤군요 🙏
죄송합니다

austinpark420 commented on src/constants/index.js:
```diff
@@ -1,10 +1,26 @@
+//     ? "http://ec2-13-125-205-115.ap-northeast-2.compute.amazonaws.com:8080"
+//     : "dev";
+
+/* 열음 서버 */
+export const BASE_SERVER_URL = "http://54.180.100.165:8080";
+
+/* 로마 서버 */
+// export const BASE_SERVER_URL =
+//   "http://ec2-13-125-205-115.ap-northeast-2.compute.amazonaws.com:8080";
+
+/* 소주캉 서버 */
+// export const BASE_SERVER_URL = "http://13.209.50.192:8080";
+
+/* 알렉스 서버 */
+// export const BASE_SERVER_URL = "http://15.164.212.121:8080";
+
+/* 필즈 서버 */
+// export const BASE_SERVER_URL = "http://15.165.204.174:8080";
```
BE 개발자분들이 각자 API 작업을 하셨으면 내려주는 데이터 형태도 조금씩 다 달랐을 것 같은데...
열음 서버를 사용하신 이유가 따로 있으신가요?

DomMorello replied:
서버 개발자분들은 전부 동일한 API 명세를 공유하기 때문에 데이터 형태가 전부 동일합니다. 동일하지 않은 경우들이 있었었는데 모두 서버 개발자분들이 합의된 API 명세를 지키지않았던 경우였고 현재 상태로는 전부 같은 API 명세를 공유합니다.

각기 다른 서버를 주석을 풀어서 연동을 해도 잘 작동하는지를 테스트하시라고 주석처리해놓았습니다. 전부 동일하게 작동해야 합니다 :)

austinpark420 commented on src/mocks/db.json:
```diff
@@ -1,210 +1,210 @@
   "products": [
     {
-      "id": 1,
+      "productId": 1,
```
id -> productId 으로 key 값이 바뀌면 생각보다 수정해야할게 많아지죠.
그래서 BE한테 목데이터를 json형태의 텍스트라도 빨리 받는게 좋은 것 같아요.

DomMorello replied:
맞습니다.. 이 부분을 수정하면서 버그가 많이 발생했었습니다 ㅎㅎㅎ

austinpark420 commented:
> 리덕스 스토어 단점
> 장바구니 목록을 redux store에서 관리하고 있습니다. 그런데 현재 기능상으로는 DB에 어차피 장바구니 목록이 저장돼있습니다. 모든 페이지를 방문할때마다 서버로 장바구니 목록을 요청해서 받아오는데 이를 또 redux store에 업데이트를 해줍니다. 이런 과정에서 어떤 기능이 수정되거나 할 때 DB와 redux store가 일관되게 유지해야 하는 번거로움이 있습니다. 이는 관리포인트가 하나 더 느는 것 같고 이렇게 함으로써 얻는 효용이 없다는 생각을 했습니다.

말씀하신대로 장바구니 목록을 store에서 관리할 필요가 없어보입니다. 현재 프로젝트에서는 유저정보만 스토어에서 관리하면 되지않을까 생각이 드네요!

austinpark420 commented:
동일한 API 호출을 2번 하는 경우가 있습니다. 확인해보시면 좋을것 같습니다.

<img width="1781" alt="Screen Shot 2022-06-10 at 8 57 47 AM" src="https://user-images.githubusercontent.com/35208965/172979305-61b2f49f-1c76-4df2-b993-38c195713708.png">

austinpark420 commented:
이메일주소를 틀리게 작성했는데 에러메세지가 비밀번호가 틀린것처럼 나오고 있네요. 이 부분 백엔드에 확인이 필요해 보입니다.
<img width="1774" alt="Screen Shot 2022-06-10 at 9 10 53 AM" src="https://user-images.githubusercontent.com/35208965/172979388-a02e90d3-81bb-4cf0-b53c-ea1891c6cac1.png">

austinpark420 commented on src/components/pages/ProductListPage/ProductCard/index.jsx:
```diff
@@ -16,33 +17,40 @@ import {
+      return;
+    }
     try {
       const response = await postBaseServerCartItem({
```
장바구니에 아이템이 담기는 경우에는 아무런 값이 네트워크탭에 아무런 값이 찍히지 않고, 담겨있는 상품을 담으려고할 때는 message가 넘어오네요.
이 부분 BE와 협의된 부분일까요? 
<img width="1782" alt="Screen Shot 2022-06-10 at 11 50 30 AM" src="https://user-images.githubusercontent.com/35208965/172980628-f0de8b1e-10a0-4c31-83a6-5eac22d32dd8.png">
<img width="1778" alt="Screen Shot 2022-06-10 at 11 50 20 AM" src="https://user-images.githubusercontent.com/35208965/172980637-5cf43bf9-c8a5-4817-96b6-f49630875b84.png">

DomMorello replied:
예! 상품이 장바구니에 추가될때는 단순히 성공 코드만 오고 실패시에만 에러메세지가 반환되도록 미리 설계된 부분입니다!

austinpark420 commented:
UX상 이미 카트에 담긴 목록의 경우 UI가 다르게 표시되면 좋을 것 같습니다.
현재는 어떤 상품이 담겨있는지 유저 입장으로서는 파악이 안되네요 ㅠ

<img width="1778" alt="Screen Shot 2022-06-10 at 11 50 20 AM" src="https://user-images.githubusercontent.com/35208965/172981039-60ac4f12-db21-4502-81bb-41808cec47b8.png">

austinpark420 commented:
UX/UI상 노란색으로 네모박스한 부분 텍스트가 변경되어야 할것 같습니다.
ex) 회원가입 -> 회원가입하기 / 로그인 -> 로그인하기 

추가로 버튼 디자인도 조금 다르면 좋을것 같습니다. 다른 서비스 참고해보시면 좋을 것 같아요!~

<img width="1767" alt="Screen Shot 2022-06-10 at 11 55 50 AM" src="https://user-images.githubusercontent.com/35208965/172981287-9828887d-42c1-41e4-9345-d1e94cf7f22e.png">
<img width="1764" alt="Screen Shot 2022-06-10 at 11 55 36 AM" src="https://user-images.githubusercontent.com/35208965/172981278-4328f5e2-08a3-4ff0-919f-65365869bd85.png">

austinpark420 reviewed (changes requested):
안녕하세요 돔하디!
BE와 같이 협업하는게 처음이라 예상치못한 이슈가 많았을 것 같은데 고생 많으셨습니다. 👍
현업에서도 BE와 커뮤니케이션을 하면서 일하시는 경우가 많을텐데 남은 미션동안 다양한 경험을 하길 바랄게요.
코드와 UX/UI 관련해서 코멘트 남겨놨으니 확인 부탁드릴게요 😄 

이제 곧 방학일텐데 잠시 리프레시하는 기간이 되면 좋겠네요!
고생 많으셨습니다.

DomMorello commented:
> 이메일주소를 틀리게 작성했는데 에러메세지가 비밀번호가 틀린것처럼 나오고 있네요. 이 부분 백엔드에 확인이 필요해 보입니다. <img alt="Screen Shot 2022-06-10 at 9 10 53 AM" width="1774" src="https://user-images.githubusercontent.com/35208965/172979388-a02e90d3-81bb-4cf0-b53c-ea1891c6cac1.png">

이 부분은 비밀번호를 5자만 입력했기 때문에 적절한 에러가 나온다고 생각합니다. 만약에 비밀번호 최소요건을 갖춰서 작성을 한다면 이메일 검증을 한 후 id 또는 비밀번호가 틀렸다는 에러메세지가 정상적으로 보여집니다.

DomMorello commented:
> 동일한 API 호출을 2번 하는 경우가 있습니다. 확인해보시면 좋을것 같습니다.
> 
> <img alt="Screen Shot 2022-06-10 at 8 57 47 AM" width="1781" src="https://user-images.githubusercontent.com/35208965/172979305-61b2f49f-1c76-4df2-b993-38c195713708.png">

React.strictmode 때문에 컴포넌트들이 두 번 렌더링되면서 fetch를 두 번 하는 것으로 파악했습니다. production에서는 strictmode가 해제되는데 현재는 배포하지않고 로컬에서 실행하다보니 develpoment 모드로 앱이 실행되고 있습니다. 그래서 strictmode가 적용된 것이고 fetch를 두 번씩 하게 됩니다. 이 부분 해결하기 위해서 strictmode를 제거했습니다.

[해당커밋](https://github.com/woowacourse/react-shopping-cart-prod/pull/38/commits/ee050e25d4124a73f1258149207b090b0b866d82)

DomMorello committed:
refactor: 두 번 렌더링 방지하기 위해 strictmode 제거

DomMorello committed:
feat: 장바구니에 담으면 사용자에게 UI로 다시 담기가 불가능하게 보여줌

장바구니 아이콘을 클릭했을 때 장바구니에 담긴 후에 disabled를 줘서
사용자가 담겼다는 것을 알 수 있도록 수정함.
ProductDetailPage 에서도 장바구니 담기 버튼을 이미 장바구니에 담겨있으면
disabled를 줘서 사용자가 이미 담긴 상품이라는 것을 알 수 있도록 함

DomMorello commented:
> 안녕하세요 오스틴님 :) 말씀해주신 내용에 대해서 코멘트 단에서 제 생각을 답변드렸습니다!

그 중에 꼭 개선하고 싶었던 부분 개선했습니다 ㅎㅎ

### strictmode 로 인한 두 번 렌더링 개선
### 장바구니 추가 UI 개선
사용자가 장바구니를 추가하면 다시 클릭하지 못하도록 UI에서 보여줌으로써 사용성을 개선했습니다!

시간을 갖고 리팩토링을 해보려고 했는데 리뷰 마감기간을 생각하지 못해서 우선 리뷰요청 보냈습니다! 피드백 주시면 감사하겠습니다! ㅎㅎㅎ

austinpark420 commented:
> 이 부분은 비밀번호를 5자만 입력했기 때문에 적절한 에러가 나온다고 생각합니다. 만약에 비밀번호 최소요건을 갖춰서 작성을 한다면 이메일 검증을 한 후 id 또는 비밀번호가 틀렸다는 에러메세지가 정상적으로 보여집니다.

앗 그렇군요.. 이부분은 UX 상 ID부터 유효성검사가 되는게 좋은 것 같습니다!

austinpark420 reviewed (approved):
안녕하세요 돔하디!~

피드백 주신 부분과 코드 확인했습니다.
간단하게 코멘트 남겨놨으니 확인부탁드릴게요.
해당 미션은 여기서 머지해도 좋을 것 같습니다. 👍 
고생많으셨고 남은 방학 잘보내세요!!
